File: lab7/lab7-2.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int leastInterval(char[] tasks, int n) {
        Map<Character, Integer> frequencyMap = new HashMap<>();
        for (char task : tasks) {
            frequencyMap.put(task, frequencyMap.getOrDefault(task, 0) + 1);
        }

        int maxFrequency = 0; 
        int maxCount = 0; 
        for (int freq : frequencyMap.values()) {
            if (freq > maxFrequency) {
                maxFrequency = freq;
                maxCount = 1; 
            } else if (freq == maxFrequency) {
                maxCount++; 
            }
        }

        int totalSlots = (maxFrequency - 1) * (n + 1) + maxCount;

        return Math.max(totalSlots, tasks.length);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        char[] tasks1 = {'A', 'A', 'A', 'B', 'B', 'B'};
        int n1 = 2;
        System.out.println(solution.leastInterval(tasks1, n1)); 

        char[] tasks2 = {'A', 'C', 'A', 'B', 'D', 'B'};
        int n2 = 1;
        System.out.println(solution.leastInterval(tasks2, n2)); 

        char[] tasks3 = {'A', 'A', 'A', 'B', 'B', 'B'};
        int n3 = 3;
        System.out.println(solution.leastInterval(tasks3, n3)); 
    }
}

File: lab7/lab7-2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_TASKS 26

int leastInterval(char* tasks, int tasksSize, int n) {
    int frequencyMap[MAX_TASKS] = {0};

    for (int i = 0; i < tasksSize; i++) {
        frequencyMap[tasks[i] - 'A']++;
    }
    int maxFrequency = 0;
    int maxCount = 0;
    for (int i = 0; i < MAX_TASKS; i++) {
        if (frequencyMap[i] > maxFrequency) {
            maxFrequency = frequencyMap[i];
            maxCount = 1;
        } else if (frequencyMap[i] == maxFrequency) {
            maxCount++;
        }
    }

    int totalSlots = (maxFrequency - 1) * (n + 1) + maxCount;

    return totalSlots > tasksSize ? totalSlots : tasksSize;
}

int main() {
    char tasks1[] = {'A', 'A', 'A', 'B', 'B', 'B'};
    int n1 = 2;
    printf("%d\n", leastInterval(tasks1, sizeof(tasks1) / sizeof(tasks1[0]), n1));

    char tasks2[] = {'A', 'C', 'A', 'B', 'D', 'B'};
    int n2 = 1;
    printf("%d\n", leastInterval(tasks2, sizeof(tasks2) / sizeof(tasks2[0]), n2));

    char tasks3[] = {'A', 'A', 'A', 'B', 'B', 'B'};
    int n3 = 3;
    printf("%d\n", leastInterval(tasks3, sizeof(tasks3) / sizeof(tasks3[0]), n3));

    return 0;
}